pkg/webhook/pod/pd_deleter.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pod
+
+import (
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	pdutil "github.com/pingcap/tidb-operator/pkg/manager/member"
+	"github.com/pingcap/tidb-operator/pkg/pdapi"
+	operatorUtils "github.com/pingcap/tidb-operator/pkg/util"
+	"github.com/pingcap/tidb-operator/pkg/webhook/util"
+	"k8s.io/api/admission/v1beta1"
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+func (pc *PodAdmissionControl) admitDeletePdPods(pod *corev1.Pod, ownerStatefulSet *apps.StatefulSet, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient) *v1beta1.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+	tcName := tc.Name
+	ordinal, err := operatorUtils.GetOrdinalFromPodName(pod.Name)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	isMember, err := IsPodInPdMembers(tc, pod, pdClient)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	isOutOfOrdinal, err := operatorUtils.IsPodOutOfOrdinal(pod, *ownerStatefulSet.Spec.Replicas)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	isUpgrading := IsStatefulSetUpgrading(ownerStatefulSet)
+
+	klog.Infof("receive delete pd pod[%s/%s] of tc[%s/%s],isMember=%v,isOutOfOrdinal=%v,isUpgrading=%v", namespace, name, namespace, tcName, isMember, isOutOfOrdinal, isUpgrading)
+
+	// each pd instance must be member deleted from tc-cluster before deleting its pod.
+	if !isMember {
```

Comment:
When a pod is not a member, it may be the pod is not registered into PD yet. You need to consider this situation. When programming in Kubernetes, you should always think carefully about the decisions you make, because of the Kubernetes eventual consistency model.

Reply:
fix